### ISSUE TITLE
DVD Folder Stacking Broken

### ISSUE THREAD

**Issue by adamulrich** [xbmc/xbmc#16109]
### Describe the bug
DVD Folder Stacking is broken.

Using the naming convention, such as "d:\movies\Gettysburg (1993) dvd1" and "d:\movies\Gettysburg (1993) dvd2", it will add these to as Videos with no information to the library (they will be listed as "video"), and will not correctly stacked.

## Expected Behavior
Here is a clear and concise description of what was expected to happen:

DVD Folder Stacking to work correctly.

## Actual Behavior

DVD Folders labeled dvd1 and dvd2 are added with placeholder metadata, and have to be manually edited, but can never be stacked.

## Possible Fix

I've looked at the code, and it appears that FileItem.StackFolders correctly identifies patterns, but it appears it does nothing does nothing with the data. 

It also appears that FileItem.StackFiles correctly handles ISO files such as d:\movies\Gettysburg (1993) dvd1.iso and d:\movies\Gettysburg (1993) dvd2.iso.

**What isn't clear** 

- **shoud StackFolders should do something with the stack, or does it expects StackFiles to do something to complete the stacking?**

- Stack Files uses regular expressions that expect a 4 part <title> <Volume> <Ignore period> <Extension>, and it uses the m_videoStackRegExps, not the m_folderStackRegExps that StackFolders uses. When it gets to the regular expression parsing, it looks at the video_ts.ifo file, and splits up the name as VI,  D, EO_TS, IFO, and not "Gettysburg (1993) 2", because it is looking at the file name, not the folder the file is in. 

It appears that if you change file1 and file2 in StackFiles to FilePath1 and FilePath2, that this could fix the problem, but it would likely break ISO stacking.

In looking through the history of this file, it looks like mkortstiege initially made significant changes to this, and would be a good candidate to know/understand what the intent of StackFolders and StackFiles is?

### To Reproduce

1. in a folder that your library has, make two copies of any dvd folder
2. rename the dvd folder as Movie (2019) dvd1 and another folder as Movie (2019) dvd2
2. start kodi and go to your movies and run update library.
3.

### Debuglog

The debuglog can be found here:

### Screenshots 
Here are some links or screenshots to help explain the problem:

## Additional context or screenshots (if appropriate)
Here is some additional context or explanation that might help:

### Your Environment
Used Operating system:

 - [x ] Android
 - [ ] iOS
 - [ ] Linux
 - [ ] OSX
 - [ ] Raspberry-Pi
 - [ x ] Windows
 - [ ] Windows UWP

 - Operating system version/name:
 - Kodi version: 18.2

*note: Once the issue is made we require you to update it with new information or Kodi versions should that be required.
Team Kodi will consider your problem report however, we will not make any promises the problem will be solved.*

**Comment by KarellenX** [comment]
I can also replicate the issue.

If memory serves me well, I think I may have come across the same issue in v17

**Comment by KarellenX** [comment]
Log here... https://paste.kodi.tv/himuzacehi.kodi
Scrape starts at Line 824

**Comment by BAzrael** [comment]
Just guessing, but may be connected to https://github.com/xbmc/xbmc/pull/15736

**Comment by SeaMonkey82** [comment]
Still broken in 18.9.

**Comment by oe3ide** [comment]
Hi, any progress on this? Thanks!

**Comment by KarellenX** [comment]
Unfortunately no progress.

**Comment by oe3ide** [comment]
ok thanks
a workaround is to name the files like the folders...

ie.
The Rock (1996)/The Rock (1996).cd1.avi
                            The Rock (1996).cd2.avi



Am Di., 24. Nov. 2020 um 00:09 Uhr schrieb KarellenX <
notifications@github.com>:

**Comment by Nehemoth** [comment]
Still broken in Kodi 19.1

:(

**Comment by KarellenX** [comment]
Yes, still broken. None of the developers have been able to look into the issue.

Mabye @arnova or @the-black-eagle might have an idea as they just fixed the stream details issue which also affected the same types of video.

**Comment by ukanuk** [comment]
VIDEO_TS / BDMV folder backups were not added to the #21685 tests it seems, I only see test files for MP4, ISO, and MKV. 

Additionally, what folder structure does Kodi prefer?
* Parts in parent folder as shared by @oe3ide or parts standalone as shared by @adamulrich? Taking a look at test files in #21685, it appears a separate parent folder is the only way it is tested, which does not match the screenshot in  [Kodi split video files docs](https://kodi.wiki/view/Naming_video_files/Movies#Split_Video_Files) but makes more sense to me and does match [Jellyfin docs](https://jellyfin.org/docs/general/server/media/shows) for what little that's worth (nice having consistency across software, but not must-have of course).

* Maybe it matters depending on whether it's separate video files, an ISO, or a VIDEO_TS / BDMV folder backup?
* Maybe the separator between parts matters? [Jellyfin docs clearly say](https://jellyfin.org/docs/general/server/media/shows) supported separators are ` (space)`, `. (period)`, `- (dash)`, and `_ (underscore)`, but I'm not seeing that noted in [Kodi split video files docs](https://kodi.wiki/view/Naming_video_files/Movies#Split_Video_Files)

```
Movies
├──PartsInFolderBluray (1991)
│  ├──PartsInFolderBluray (1991) disc1
│  │  ├──BDMV
│  │  └──CERTIFICATE
│  └──PartsInFolderBluray (1991) disc2
│     ├──BDMV
│     └──CERTIFICATE
├──PartsInFolderDvd (1992)
│  ├──PartsInFolderDvd (1992) disc1
│  |  ├──AUDIO_TS
│  |  └──VIDEO_TS
│  └──PartsInFolderDvd (1992) disc2
│     ├──AUDIO_TS
│     └──VIDEO_TS
├──PartsInFolderIso (1993)
│  ├──PartsInFolderIso (1993) disc1.iso
│  └──PartsInFolderIso (1993) disc2.iso
├──PartsInFolderMp4 (1994)
│  ├──PartsInFolderMp4 (1994).cd1.mp4
│  └──PartsInFolderMp4 (1994).cd2.mp4
│
├──PartsStandaloneBluray (1995) disk1
│  ├──BDMV
│  └──CERTIFICATE
├──PartsStandaloneBluray (1995) disk2
├  ├──BDMV
├  └──CERTIFICATE
├──PartsStandaloneDvd (1996) disk1
│  ├──AUDIO_TS
│  └──VIDEO_TS
├──PartsStandaloneDvd (1996) disk2
│  ├──AUDIO_TS
│  └──VIDEO_TS
├──PartsStandaloneIso (1997)_disk1.iso
├──PartsStandaloneIso (1997)_disk2.iso
├──PartsStandaloneMp4 (1998).cd1.mp4
└──PartsStandaloneMp4 (1998).cd2.mp4
```

For extra context, I prefer disc folder backups (VIDEO_TS / BDMV) for now and am wondering trying to figure out the best way to organize these collections I own (including probably needing NFO files, but that's probably another topic):
* 2-disc DVD [Monty Python and the Holy Grail (1975)](https://www.barnesandnoble.com/w/dvd-monty-python-and-the-holy-grail-graham-chapman/3626049), seems simple enough with movie on disc 1 and bonus content on disc 2
* 5-disc Bluray [Rush Hour Trilogy](https://www.blu-ray.com/movies/Rush-Hour-Trilogy-Blu-ray/164037/), slightly more complex with three movies on three discs and two discs with bonus content.
* 15-disc / 28-film Bluray [Abbott & Costello: The Complete Universal Pictures Collection](https://shoutfactory.com/products/abbott-costello-the-complete-universal-pictures-collection-80th-anniversary-edition?variant=40080693493894), fourteen discs each with two movies and movie-specific extras and a fifteenth disc with general Abbott & Costello extras.

This is my best guess for how Kodi would prefer that organized, although it is not currently scanning correctly. I would guess an [NFO multi-episode file](https://kodi.wiki/view/NFO_files/Episodes#Multi-Episode_Files) is necessary with some combination of [Add-on:Extras](https://kodi.wiki/view/Add-on:Extras), [<folderstacking> advancedsettings.xml settings](https://kodi.wiki/view/Advancedsettings.xml#folderstacking), and [Add-on:Special Features](https://kodi.wiki/view/Add-on:Special_Features) or this [forum workaround](https://forum.kodi.tv/showthread.php?tid=327042&pid=3113106#pid3113106).

```
Movies
├──Abbott & Costello: The Complete Universal Pictures Collection
│  ├──Abbott & Costello The Complete Universal Pictures Collection disc01
│  │  ├──BDMV
│  │  │  └──index.nfo
│  │  └──CERTIFICATE
│  ├─ 
│  ├─ ⋮
│  ├─ 
│  └──Abbott & Costello The Complete Universal Pictures Collection disc15
│     ├──BDMV
│     │  └──index.nfo
│     └──CERTIFICATE
├──Monty Python and the Holy Grail (1975)
│  ├──Monty Python and the Holy Grail (1975) disc1
│  |  ├──AUDIO_TS
│  |  └──VIDEO_TS
│  │     └──VIDEO_TS.nfo
│  └──Monty Python and the Holy Grail (1975) disc2
│     ├──AUDIO_TS
│     └──VIDEO_TS
│        └──VIDEO_TS.nfo
├──Rush Hour (1998)
│  └──Rush Hour (1998)
│     ├──BDMV
│     │  └──index.nfo
│     └──CERTIFICATE
├──Rush Hour 2 (2001)
│  └──Rush Hour 2 (2001)
│     ├──BDMV
│     │  └──index.nfo
│     └──CERTIFICATE
├──Rush Hour 3 (2007)
│  └──Rush Hour 3 (2007)
│     ├──BDMV
│     │  └──index.nfo
│     └──CERTIFICATE
└──Rush Hour Trilogy Bonus
   ├──Rush Hour Trilogy disc4
   │  ├──BDMV
   │  │  └──index.nfo
   │  └──CERTIFICATE
   └──Rush Hour Trilogy disc5
      ├──BDMV
      │  └──index.nfo
      └──CERTIFICATE
```

**Comment by szb** [comment]
@ukanuk Did you happen to make any progress on your questions since you posted? I am in the same boat as you I presume. I have a large physical media collection that I want to digitize so I can play the backups directly. With disc rot and other issues looming, it would be great if there was a way to play the original files well into the future. I'll poke around the links you posted to see if I can make any progress.

**Comment by ukanuk** [comment]
> Edit - - for shows, I had to use format SxxExxEyyEzz for automatic correct scanning rather than SxxExx-Ezz. This is shown at https://kodi.wiki/view/Naming_video_files/Episodes.

Still not sure what to do about the Abbott and Costello collection.

**Comment by KarellenX** [comment]
@78andyp 
Just an fyi in the event you make changes to dvd's

**Comment by 78andyp** [comment]
@KarellenX - thanks - looking at @ukank's last post.
Proper episode recognition for blurays is what we are currently testing on slack.
SxxEaa-bb is here with #26439. It's just waiting a final check by @CrystalP 
Two movies on one bluray disc will be here after episodes are sorted - it will be a sort of extension to #26277.
Stacks for folders just needs tidying up - the bulk of it is in #25487.
Then dvd vfs needs adding - that's already 80% done as well but will need modifying in light of some of the bluray changes.

**Comment by KarellenX** [comment]
That is great news! Thanks @78andyp 
I don't use bluray or dvd folders as I rip my media to mkv files, so have only been loosely following conversation.

**Comment by adutta98** [comment]
I wish this could be fixed, is it still being worked upon? Movies like Avataar & sequals, LotR, Gandhi, The Lawrence of Arabia have been released on multiple UHD's, and I know a bunch more use cases, would be nice to see this fixed. Appreciate all the hard work.

**Comment by 78andyp** [comment]
Yes absolutely. Once I've got all the bluray stuff in, this is next.

**Comment by adutta98** [comment]
@78andyp Thank you Sir!

**Comment by grendizerus** [comment]
Hello, there is a problem on stacked AVI files as well : the length of video 2 is not added to the length of video 1. This means , for a 2 hours movie, the progress bar shows a length of 1 hour. The movie plays fine but if I stop it after the first hour and resume the playback, I can't, Kodi thinks the movie is fully watched and does not offer me to resume.
I tried oe3ide's workaround and it works. So for example I renamed my 2 parts movie as such:
\Insomnia (2002)\Insomnia (2002).cd1.avi
\Insomnia (2002)\Insomnia (2002).cd2.avi
This works fine. But the Kodi WIKI offers more options for renaming multi parts movies.
However, Thexai said the bug is now fixed. Is this stacked AVI problem fixed in Kodi 22 "Piers"?
Thanks
DS

**Comment by 78andyp** [comment]
The parts need to be named - yes.
These are the regexes:
```
      "(.*?)([ _.-]*(?:cd|dvd|p(?:(?:ar)?t)|dis[ck]|file)[ _.-]*[0-9]+)(.*?)(\\.[^.]+)$",
      "(.*?)([ _.-]*(?:cd|dvd|p(?:(?:ar)?t)|dis[ck])[ _.-]*[a-h])(.*?)(\\.[^.]+)$",
      "^(.+)((?:[ ._-]|(?<=\\)))[a-h])()(\\.[^.]+)$",
```

So anything that ends with cd/dvd/part/disc/disk followed by 0-9 or a-h will be stacked.
In your case cd1/cd2 but you could also have 'part 1', 'part 2' etc..

**Comment by grendizerus** [comment]
Any syntax different than oe3ide's workaround will not work in Kodi v21. These files used to be named : 
Insomnia (2002)\Insomnia (2002) Audio French Part 01.avi
When I tried to play them in Kodi 21, the movie length problem appeared. I tried the same with other multi parts movie and got the same issues, that were ONLY fixed by oe3ide's syntax. 
There is a thread about this on the Kodi forum : 
[https://forum.kodi.tv/showthread.php?tid=365890](url)
Could anyone try in Kodi 22 to check if the problem is still here? Thanks
DS

**Comment by grendizerus** [comment]
OK after more tests I found out there was a mistake on my side. There should be no space between "part" or "cd" or "dvd" and the part number, so the proper syntax on my original files should have been:
Insomnia (2002) Audio French Part01.avi
Sorry for the trouble :)

**Comment by 78andyp** [comment]
A space should be allowed just looking at the regex. Let me try this end.

**Comment by KarellenX** [comment]
Hmm, I don't know about allowing a space for stacking.
There are users that do want stacking, and use the non-space naming. But there are also times where you want the files numbered but not stacked.
Sorry, I don't have an example to give, just my experience on the forum.

**Comment by 78andyp** [comment]
Odd. I've just tried it here:
<img width="551" height="316" alt="Image" src="https://github.com/user-attachments/assets/62a25d57-bb7f-4fcc-aad3-6cd8d3793abe" />
And it works.
But as @grendizerus has got it working and I'm assuming we've not had complaints so far @KarellenX  - I'll just do nothing and leave it alone!!

**Comment by grendizerus** [comment]
Thanks for testing 78andyp. Could you try : The Hobbit Part 01.mp4 ?
Are you on Kodi 21 or 22?

**Comment by 78andyp** [comment]
v22
Using Part 01 and Part 02 works fine here.

**Comment by KarellenX** [comment]
Thanks @78andyp
